### PR TITLE
Add record rosbag launch

### PR DESCRIPTION
**Pull request by jaagut**
## Proposed changes

- Add `record_bag.launch`
    - Add `sim` parameter
    - Also throttle pointclouds
- Rename `tools.launch` to `monitoring.launch`
- Include this in `monitoring.launch`

TODOs:
- [ ] Evaluate whether image sub-sampling aligns with head-behavior -> only one perspective?

## Related Issues
#223
https://github.com/bit-bots/wolfgang_robot/pull/271

### CODE REVIEW

**Comment by jaagut** [comment]
[Here](https://github.com/bit-bots/wolfgang_robot/blob/895e71785d436830b7e9b89d9ac18ec6570d84a0/wolfgang_robocup_api/launch/record_bag.launch) a similar launch already exists…

We need to compare these two and decide how to move forward.

**Comment by jaagut** [comment]
This has to be tested and merged together with:
bit-bots/wolfgang_robot/pull/271

**Comment by HaukeSchnau** [comment]
😢✝️

`<executable cmd="ros2 bag record --all -o rosbag_test_$(command 'bash -c &quot;date +%s |tr -d \\&quot;\\n\\&quot;&quot;')"/>`

File: bitbots_utils/launch/record_bag.launch [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    <!-- The command is a bit complicated, because we want to have a unique filename
+         and the date command would normally include a newline at the end. -->
+    <!-- We also set include-hidden-topics twice, because otherwise, the else case would produce some unprintable character that breaks the command. -->
+    <executable cmd="ros2 bag record \
+    -o $(env HOME)/rosbags/ID_$(env ROBOCUP_ROBOT_ID $(env ROBOT_NAME unknown_robot))_$(command 'bash -c &quot;date +%s |tr -d \\&quot;\\n\\&quot;&quot;') \
+    --node-name ros2_bag_record \
+    --include-hidden-topics \
+    --include-unpublished-topics \
+    --polling-interval 1000 \
+    $(eval '\'--use_sim_time\' if \'$(var sim)\'==\'true\' else \'--include-hidden-topics\'') \
```
Comment (timonegk) [inline bitbots_utils/launch/record_bag.launch:21]: This is really horrible and unmaintainable. ROS 2 added Python launch files. They are specifically made to deal with more complicated logic. If you don't want to learn how to write them, I can do it. But I would really prefer if we don't merge this.

File: bitbots_utils/launch/tools.launch [hunk 2]
```diff
@@ -1,12 +1,16 @@
 <?xml version="1.0"?>
 <launch>
     <arg name="monitoring" default="true" description="true: start system monitor and udp bridge" />
+    <arg name="recording" default="true" description="true: start the ros bag recording" />
```
Comment (timonegk) [inline bitbots_utils/launch/tools.launch:4]: In `robocup_teamplayer.launch`, this argument is called `record`. I would prefer to keep the name consistent across launch files.

**Review by timonegk — CHANGES_REQUESTED** [review]

File: bitbots_utils/launch/record_bag.launch (hunk 1 above)
Comment (jaagut) [reply bitbots_utils/launch/record_bag.launch:21]: ChatGPT converted it to this believable code. Will test it soon...
```py
from launch import LaunchDescription
from launch.actions import DeclareLaunchArgument, ExecuteProcess
from launch.substitutions import EnvironmentVariable, LaunchConfiguration, Command
from launch_ros.actions import Node

def generate_launch_description():
    return LaunchDescription([
        DeclareLaunchArgument(
            'sim', default_value='false', description='true: Use simulation time'),
        DeclareLaunchArgument(
            'max_image_frequency', default_value='1.0', description='Max frequency [hz] for recording images'),
        DeclareLaunchArgument(
            'max_pointcloud_frequency', default_value='1.0', description='Max frequency [hz] for recording pointclouds'),

        Node(
            package='topic_tools', executable='throttle', output='screen',
            name='record_rosbag_drop_images',
            arguments=[
                'messages', '/camera/image_proc',
                LaunchConfiguration('max_image_frequency'),
                '/camera/image_to_record'
            ]
        ),
        Node(
            package='topic_tools', executable='throttle', output='screen',
            name='record_rosbag_drop_pointclouds',
            arguments=[
                'messages', '/line_mask_relative_pc',
                LaunchConfiguration('max_pointcloud_frequency'),
                '/line_mask_relative_pc_to_record'
            ]
        ),

        ExecuteProcess(
            cmd=[
                'ros2', 'bag', 'record',
                '-o', EnvironmentVariable('HOME') + '/rosbags/ID_' +
                EnvironmentVariable('ROBOCUP_ROBOT_ID', default_value=EnvironmentVariable('ROBOT_NAME', default_value='unknown_robot')) +
                '_' + Command(['bash', '-c', '"date +%s |tr -d \\"\\n\\"']),
                '--node-name', 'ros2_bag_record',
                '--include-hidden-topics',
                '--include-unpublished-topics',
                '--polling-interval', '1000',
                '--use_sim_time' if LaunchConfiguration('sim')=='true' else '--include-hidden-topics',
                '/animation',
                '/audio/audio_info',
                '/audio/audio',
                '/ball_obstacle_active',
                '/ball_position_relative_filtered',
                '/ball_relative_filtered',
                '/ball_relative_movement',
                '/balls_relative',
                '/camera/camera_info',
                '/camera/image_to_record',
                '/clock',
                '/cmd_vel',
                '/cop_l',
                '/cop_r',
                '/core/power_switch_status',
                '/debug_markers',
                '/debug/approach_point',
                '/debug/ball_twist',
                '/debug/dsd/body_behavior',
                '/debug/dsd/hcm',
                '/debug/dsd/head_behavior',
                '/debug/dsd/localization',
                '/debug/used_ball',
                '/debug/which_ball_is_used',
                '/diagnostics',
                '/diagnostics_agg',
                '/DynamixelController/command',
                '/field_boundary_relative',
                '/game_controller_connected',
                '/gamestate',
                '/goal_pose',
                '/head_mode',
                '/imu_head/data',
                '/imu/data_raw',
                '/joint_states',
                '/line_mask_relative_pc_to_record',
                '/motion_odometry',
                '/move_base/current_goal',
                '/obstacles_relative',
                '/pose_with_covariance',
                '/robot_state',
                '/robots_relative',
                '/robots_relative_filtered',
                '/rosout',
                '/server_time_clock',
                '/speak',
                '/strategy',
                '/system_workload',
                '/team_data',
                '/tf',
                '/tf_static',
                '/time_to_ball'
            ],
            output='screen'
        ),
    ])
```

Comment (HaukeSchnau) [reply bitbots_utils/launch/record_bag.launch:21]: Can we now change the (still quite hacky) `Command(['bash', '-c', '"date +%s |tr -d \\"\\n\\"'])` to a simple call to something like python's `time`? 

Also, I'm curious if we can now clean up`'--use_sim_time' if LaunchConfiguration('sim')=='true' else '--include-hidden-topics'` by completely omitting the parameter if the condition is not met. (Previously we had to resort  to an empty string which caused problems)

**Comment by jaagut** [comment]
Migrating this XML launch file to a working and complete Python launch file was more difficult than expected. The progress is being tracked in #223.

As this currently works, we want to use this XML launch file meanwhile.

**Review by timonegk — APPROVED** [review]

**Pull request merged**